src/java.base/windows/classes/sun/nio/ch/WEPollSelectorImpl.java

```diff
@@ -0,0 +1,276 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.nio.ch;
+
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedSelectorException;
+import java.nio.channels.Pipe;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.spi.SelectorProvider;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static sun.nio.ch.WEPoll.*;
+
+/**
+ * Windows wepoll based Selector implementation
+ */
+class WEPollSelectorImpl extends SelectorImpl {
+    // maximum number of events to poll in one call to epoll_wait
+    private static final int NUM_EPOLLEVENTS = 256;
+
+    // wepoll handle
+    private final long eph;
+
+    // address of epoll_event array when polling with epoll_wait
+    private final long pollArrayAddress;
+
+    // maps SOCKET to selection key, synchronize on selector
+    private final Map<Integer, SelectionKeyImpl> fdToKey = new HashMap<>();
+
+    // pending new registrations/updates, queued by setEventOps
+    private final Object updateLock = new Object();
+    private final Deque<SelectionKeyImpl> updateKeys = new ArrayDeque<>();
+
+    // interrupt/wakeup
+    private final Object interruptLock = new Object();
+    private boolean interruptTriggered;
+    private final PipeImpl pipe;
+    private final int fd0Val, fd1Val;
+
+    WEPollSelectorImpl(SelectorProvider sp) throws IOException {
+        super(sp);
+
+        this.eph = WEPoll.create();
+        this.pollArrayAddress = WEPoll.allocatePollArray(NUM_EPOLLEVENTS);
+
+        // wakeup support
+        try {
+            this.pipe = new PipeImpl(sp, /*buffering*/ false);
+        } catch (IOException ioe) {
+            WEPoll.freePollArray(pollArrayAddress);
+            WEPoll.close(eph);
+            throw ioe;
+        }
+        this.fd0Val = pipe.source().getFDVal();
+        this.fd1Val = pipe.sink().getFDVal();
+
+        // register one end of the pipe for wakeups
+        WEPoll.ctl(eph, EPOLL_CTL_ADD, fd0Val, WEPoll.EPOLLIN);
+    }
+
+    private void ensureOpen() {
+        if (!isOpen())
+            throw new ClosedSelectorException();
+    }
+
+    @Override
+    protected int doSelect(Consumer<SelectionKey> action, long timeout)
+        throws IOException
+    {
+        assert Thread.holdsLock(this);
+
+        // epoll_wait timeout is int
+        int to = (int) Math.min(timeout, Integer.MAX_VALUE);
+        boolean blocking = (to != 0);
+
+        int numEntries;
+        processUpdateQueue();
+        processDeregisterQueue();
+        try {
+            begin(blocking);
+            numEntries = WEPoll.wait(eph, pollArrayAddress, NUM_EPOLLEVENTS, (int) timeout);
```

Comment:
I think  you meant to use 'to' in place of '(int) timeout'.

Reply:
Yes, thanks, this crept it when re-basing the change for the main line.